### PR TITLE
[data] Fix `move_tensors_to_device` utility for the list/tuple[tensor] case

### LINKED ISSUE #52989: Release test training_ingest_benchmark-task=image_classification.skip_training.parquet.fault_tolerance failed
Release test **training_ingest_benchmark-task=image_classification.skip_training.parquet.fault_tolerance** failed. See https://buildkite.com/ray-project/release/builds/41525#0196cd76-e6f9-4f81-a902-b4f274f23ccb for more details.

Managed by OSS Test Policy

### PR DESCRIPTION
**Pull request by justinvyu**
## Summary

Fix moving a list/tuple of tensors to device. Previously, the utility tried concatenating the sequence, rather than just moving each tensor and returning the list. Ex: the correct behavior should be: `(images, labels) -> (images.to(device), labels.to(device))`.

This PR also does some cleanup to make the type-handling logic more modular and adds some doctests + unit tests.

Fixes the `training_ingest_benchmark-task=image_classification.skip_training.parquet.fault_tolerance` release test failure, which tests the CPU case where `combine_chunks=True`, and the collate function returns a tuple, which leads to the `tuple[tensor]` case.

## Related Issues

Closes https://github.com/ray-project/ray/issues/52989

### CODE REVIEW

File: python/ray/air/tests/test_torch_tensor_utils.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
+import sys
+
+import pytest
+import torch
+
+from ray.air._internal.torch_utils import move_tensors_to_device
+
+
+def test_move_tensors_to_device():
```
Comment (srinathk10) [inline python/ray/air/tests/test_torch_tensor_utils.py:9]: Nice thanks for fixing the behavior for Tuple of Tensors. Also look at test_iter_torch_batches_gpu.py which has coverage for all the types expected in `move_tensors_to_device`

**Review by srinathk10 — APPROVED** [review]

File: python/ray/air/tests/test_torch_tensor_utils.py (hunk 1 above)
Comment (justinvyu) [reply python/ray/air/tests/test_torch_tensor_utils.py:9]: Got it, fixed the test cases there for lists/tuple types. Btw, that test does not cover currently the nested lists/tuples cases since that would only arise with multiple blocks of data which doesn't happen in the toy dataset.

I think we can keep both tests since that one is testing e2e usage of iter_torch_batches with all the possible combinations of batch type, supported collate output, etc.

This test is a unittest on the `move_tensors_to_device` method.

File: python/ray/air/_internal/torch_utils.py [hunk 2]
```diff
@@ -432,6 +464,12 @@ def move_tensors_to_device(
 ) -> TensorBatchReturnType:
     """Move tensors to the specified device.
 
+    Concatenate nested lists/tuples of tensors along the first (batch) dimension.
+    For example, for the input
+    ((feature_0_chunk_0,), (feature_1_chunk_0, feature_1_chunk_1))
+    the output will be (feature_0_chunk_0, feature_1_chunk_0+1)
+    where each feature is concatenated along the batch dimension.
```
[the remark this answers is not in the document]
Comment (justinvyu) [reply python/ray/air/_internal/torch_utils.py:471]: This is the optimization done by https://github.com/ray-project/ray/pull/52548.

Batches are represented by Arrow tables, where each column of the table is represented by a `ChunkedArray` which is basically a bunch of slices of rows to form a single batch. The typical case where data is not contiguous is at the block boundaries, or if the batch size is really large and spans multiple blocks.

Screenshot 2025-05-20 at 3 26 41 PM

We used to always combine chunks before moving to GPU which involves making a CPU copy of the batch. Now, we keep the batch in chunks (`List/Tuple[Tensor]`) and copy each chunk to the GPU.

Now, `List/Tuple[Tensor]` can represent a chunked version of a single `Tensor`. But I think it's getting conflated with an actual `List/Tuple[Tensor]` returned by a collate function. We may want to add a lightweight wrapper class to distinguish the 2 cases.

**Review by matthewdeng — APPROVED** [review]
